Implement FxCop CA1307 Specify StringComparison

basoundr opened the pull request:
This change implements the FxCop rule CA1307 to pick the overload with
StringComparison if possible.

basoundr commented:
Tagging @dotnet/roslyn-analysis

srivatsn commented on src/Analyzer.Utilities/Extensions/IEnumerableOfIMethodSymbolExtensions.cs:
```diff
@@ -0,0 +1,70 @@
+            });
+        }
+
+        public static IEnumerable<IMethodSymbol> GetMethodOverloadsWithDesiredParameterAtTrailing(
```
Add doc comments please.

basoundr replied:
Added comments

basoundr commented:
@dotnet-bot retest Windows Release please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers/job/master/job/windows_release_prtest/88/
// Retest reason: Nuget failure

basoundr commented:
@dotnet/roslyn-analysis good to checkin?

mavasani commented on src/Analyzer.Utilities/Extensions/IEnumerableOfIMethodSymbolExtensions.cs:
```diff
@@ -0,0 +1,92 @@
+    public static class IEnumerableOfIMethodSymbolExtensions
+    {
+        /// <summary>
+        /// Returns a list of method symbols from a given list of the method symbols, which has its parameter type as
```
Seems like a very specific helper. Are you planning to use this for any future analyzers? Otherwise, I would recommend moving this into the analyzer file.

basoundr replied:
Yes, this is used in the analyzer `SpecifyIFormatProvider.cs` for which I have a PR out #912. This helper will be shared between them

mavasani commented on src/Analyzer.Utilities/Extensions/IEnumerableOfIMethodSymbolExtensions.cs:
```diff
@@ -0,0 +1,92 @@
+        /// </summary>
+        /// <param name="methods">List of <see cref="IMethodSymbol"/> to scan for possible overloads</param>
+        /// <param name="selectedOverload"><see cref="IMethodSymbol"/> that is currently picked by the user</param>
+        /// <param name="expectedParameterType"><see cref="INamedTypeSymbol"/> type of the leading parameter or the trailing parameter</param>
```
expectedTrailingParameterType? Also update the doc comment and use named argument for argument `true` in the below call.

mavasani commented on src/Analyzer.Utilities/Extensions/IEnumerableOfIMethodSymbolExtensions.cs:
```diff
@@ -0,0 +1,92 @@
+        /// <returns></returns>
+        public static IMethodSymbol GetSingleOrDefaultMemberWithName(this IEnumerable<IMethodSymbol> stringFormatMembers, string displayName)
+        {
+             return stringFormatMembers?.Where(member => string.Equals(member.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat), displayName, StringComparison.OrdinalIgnoreCase)).SingleOrDefault();
```
This uses a language specific format, please move this helper to the analyzer file.

basoundr replied:
@mavasani  This was discussed in the earlier PR as well. Though the format used here is named CSharpErrorMessageFormat, this has got nothing to do with CSharp. This is just one other SymbolDisplayFormat and it is used by CSharp. The reason to use this instead of the each symbol's default display format is, the default display for VB symbols are `Public Property Name as String` which doesn't fit well in the error messages.

Both the base type analyzer and the helper is in language agnostic part of the code so I feel that there is no harm in keeping it here.

mavasani commented on src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs:
```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+            {
+                var stringComparisonType = csaContext.Compilation.GetTypeByMetadataName("System.StringComparison");
+                var stringType = csaContext.Compilation.GetSpecialType(SpecialType.System_String);
+                var currentCulture = stringComparisonType.GetMembers("CurrentCulture").SingleOrDefault();
```
Using `SingleOrDefault` will cause an exception if there are multiple definitions found (for example a custom build mscorlib or a user defined type in source which defines multiple members with same name). Its generally better to use `.OfType<IXXXSymbol>().FirstOrDefault()`.

basoundr replied:
The inclusion to have `.OfType<IxxxSymbol>` sounds good to me but `SingleOrDefault()` is better pick here because we are searching for predefined mscorlib members and use those members to display the error message. Sure, with `FirstOrDefault()` there will not be a crash but the behavior will change we will not notice this. With SingleOrDefault(), we will catch if something changes in these predefined types

mavasani commented on src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs:
```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+                }
+
+                var stringCompareToNamedMethods = stringType.GetMembers("CompareTo").OfType<IMethodSymbol>();
+                var stringCompareToParameterString = stringCompareToNamedMethods.GetSingleOrDefaultMemberWithName("string.CompareTo(string)");
```
I don't think we want to encourage get members by their signature string - it should be easy to fetch them by getting the type symbols for parameters and then iterating the overloads to find the matching member.

basoundr replied:
Took it offline. Moving `GetSingleOrDefaultMemberWithName` out of the helper so that the other users dont start using without understanding the implication of using the Display format to compare members, which is acceptable in this case

mavasani commented on src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs:
```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+                var boolString = csaContext.Compilation.Language == LanguageNames.CSharp ? "bool" : "Boolean";
+                var intString = csaContext.Compilation.Language == LanguageNames.CSharp ? "int" : "Integer";
+                var stringCompareNamedMethods = stringType.GetMembers("Compare").OfType<IMethodSymbol>();
+                var stringCompareParameterStringStringBool = stringCompareNamedMethods.GetSingleOrDefaultMemberWithName($"string.Compare(string, string, {boolString})");
```
Please change all this code to be based on finding the overload based on given list of special types, no need for any string matching here.

mavasani commented on src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs:
```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+                overloadMap.AddKeyValueIfNotNull(stringCompareToParameterObject, stringCompareParameterStringStringComparison);
+                overloadMap.AddKeyValueIfNotNull(stringCompareParameterStringStringBool, stringCompareParameterStringStringComparison);
+                overloadMap.AddKeyValueIfNotNull(stringCompareParameterStringIntStringIntIntBool, stringCompareParameterStringIntStringIntIntComparison);
+
```
Can you bail out if overloadMap is empty at the end?

basoundr replied:
There are other ways to generate diagnostics, so we should not bail. This `overloadMap` is for some predefined member behavior. We still need to do the check for generic behavior

mavasani commented on src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs:
```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+                    var targetMethod = invocationExpression.TargetMethod;
+
+                    if (targetMethod.IsGenericMethod ||
+                        targetMethod.ContainingType == null)
```
ContaingType cannot be null

mavasani replied:
Can you add another check to bail out early: `!targetMethod.Name.StartsWith("Compare", StringComparison.OrdinalIgnoreCase)`

mavasani replied:
Ah, I see that the logic below is not specific to String.CompareXXX methods, ignore my last comment.

mavasani commented on src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs:
```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+
+                    if (overloadMap.Count != 0 && overloadMap.ContainsKey(targetMethod))
+                    {
+                        ReportDiagnostic(
```
Should we return after reporting diagnostic here?

basoundr replied:
yes

basoundr commented:
@mavasani Addressed your comments

mavasani commented:
:+1:

basoundr commented:
@srivatsn @mavasani I have modified the diagnostic message for CA1305 and CA1307 to be shorter and also introduced some helpers. Please review this change :)

mavasani commented on src/System.Runtime.Analyzers/Core/SystemRuntimeAnalyzersResources.resx:
```diff
@@ -331,10 +331,10 @@
   </data>
   <data name="SpecifyStringComparisonMessage" xml:space="preserve">
-    <value>Specify StringComparison</value>
+    <value>Because the behavior of '{0}' could vary based on the current user's locale settings, replace this call in '{1}' with a call to '{2}'.</value>
```
I like the new message much better 👍

mavasani replied:
Minor grammar suggestion: avoid starting the sentence with "Because". You might want to just remove the term "because" in the diagnostic.

mavasani commented:
👍
